From 14da1728e0348f5968e3d58e454c9fb4ddaa405e Mon Sep 17 00:00:00 2001
From: Steve Meek <stevemeek82@gmail.com>
Date: Tue, 13 Oct 2026 17:35:03 +0100
Subject: [PATCH] umbim: register the network's DNS servers as peer DNS

Since 03e0a608e0 ("package: utils: umbim: add dns servers independent
of peerdns") the DNS block on the wwan_4/wwan_6 interfaces only runs
when dhcp/dhcpv6 is 1. Both default to "auto", so in the default setup
an MBIM connection comes up with no DNS servers: the internet works but
nothing resolves.

Those servers were added to the static child interfaces as configured
"dns", which netifd's peerdns does not filter. Add them to the parent
interface with proto_add_dns_server instead, as qmi.sh does, so netifd
handles peerdns itself: by default they reach resolv.conf.auto, and with
peerdns 0 they stay out of it but are still listed in ifstatus, under
"inactive".

Tested on a Quectel RM500Q-AE (PCIe, mhi_wwan_mbim) on current main:
- before: no dns-server on wwan or wwan_4, resolv.conf.auto empty
- peerdns default: the network's two servers on wwan and in
  resolv.conf.auto, names resolve
- peerdns 0: resolv.conf.auto empty, both servers under "inactive" in
  ifstatus wwan

Signed-off-by: Steve Meek <stevemeek82@gmail.com>
---
 package/network/utils/umbim/Makefile          |  2 +-
 .../umbim/files/lib/netifd/proto/mbim.sh      | 26 +++++++------------
 2 files changed, 11 insertions(+), 17 deletions(-)

diff --git a/package/network/utils/umbim/Makefile b/package/network/utils/umbim/Makefile
index 10896f4970204c..b16681e6895441 100644
--- a/package/network/utils/umbim/Makefile
+++ b/package/network/utils/umbim/Makefile
@@ -1,7 +1,7 @@
 include $(TOPDIR)/rules.mk
 
 PKG_NAME:=umbim
-PKG_RELEASE:=1
+PKG_RELEASE:=2
 
 PKG_SOURCE_PROTO:=git
 PKG_SOURCE_URL=$(PROJECT_GIT)/project/umbim.git
diff --git a/package/network/utils/umbim/files/lib/netifd/proto/mbim.sh b/package/network/utils/umbim/files/lib/netifd/proto/mbim.sh
index 2553ada406397b..f2b79e1f226461 100755
--- a/package/network/utils/umbim/files/lib/netifd/proto/mbim.sh
+++ b/package/network/utils/umbim/files/lib/netifd/proto/mbim.sh
@@ -228,6 +228,16 @@ _proto_mbim_setup() {
 	tid=$((tid + 1))
 
 	proto_init_update "$ifname" 1
+	[ "$iptype" != "ipv6" ] && {
+		for server in $(_proto_mbim_get_field ipv4dnsserver "$mbimconfig"); do
+			proto_add_dns_server "$server"
+		done
+	}
+	[ "$iptype" != "ipv4" ] && {
+		for server in $(_proto_mbim_get_field ipv6dnsserver "$mbimconfig"); do
+			proto_add_dns_server "$server"
+		done
+	}
 	proto_send_update "$interface"
 
 	[ -z "$dhcp" ] && dhcp="auto"
@@ -253,14 +263,6 @@ _proto_mbim_setup() {
 			json_add_string proto "dhcp"
 		fi
 
-		[ "$dhcp" != 1 ] || {
-			json_add_array dns
-			for server in $(_proto_mbim_get_field ipv4dnsserver "$mbimconfig"); do
-				json_add_string "" "$server"
-			done
-			json_close_array
-		}
-
 		proto_add_dynamic_defaults
 		[ -n "$zone" ] && json_add_string zone "$zone"
 		[ -n "$ip4table" ] && json_add_string ip4table "$ip4table"
@@ -298,14 +300,6 @@ _proto_mbim_setup() {
 			[ "$sourcefilter" = "0" ] && json_add_boolean sourcefilter "0"
 		fi
 
-		[ "$dhcpv6" != 1 ] || {
-			json_add_array dns
-			for server in $(_proto_mbim_get_field ipv6dnsserver "$mbimconfig"); do
-				json_add_string "" "$server"
-			done
-			json_close_array
-		}
-
 		proto_add_dynamic_defaults
 		[ -n "$zone" ] && json_add_string zone "$zone"
 		[ -n "$ip6table" ] && json_add_string ip6table "$ip6table"
